vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java

```diff
@@ -101,4 +109,66 @@ public void setAllRowsVisibleProperty_isHeightByRowsAndIsAllRowsVisibleWork() {
         Assert.assertTrue(grid.isHeightByRows());
         Assert.assertTrue(grid.isAllRowsVisible());
     }
+
+    @Test
+    public void dataProviderListeners_gridAttached_oldDataProviderListenerRemoved() {
+        AtomicInteger listenersCount = new AtomicInteger(0);
+        AtomicBoolean oldListenerRemoved = new AtomicBoolean();
+
+        // given
+        Grid<String> grid = new Grid<>();
+        ListDataProvider<String> dataProvider = new ListDataProvider<String>(
+                Collections.emptyList()) {
+            @Override
+            public Registration addDataProviderListener(
+                    DataProviderListener<String> listener) {
+                listenersCount.incrementAndGet();
+                Registration registration = super.addDataProviderListener(
+                        listener);
+                // the first listener is added by Grid in 'setDataProvider'
+                // and it is the one we want to be removed.
+                return listenersCount.get() == 1
+                        ? Registration.combine(registration,
+                                () -> oldListenerRemoved.set(true))
+                        : registration;
+            }
+        };
+
+        // when
+        grid.setDataProvider(dataProvider);
```

Comment:
The logic feels off to me that the listeners are added
a) when the component is attached - OK
b) when the data provider is added - I would expect that here it is only added if the component was attached to begin with

Reply:
Basically, I agree that in most cases it's a non sense to call `refreshAll` or `refreshItem` on detached component, because the updated won't be seen anyway. But there are a couple of concerns to do that:
1. Some 'exotic' cases when `refresh` methods are called even before component attach and some logic relies on that.
2. Should we distinguish between 'initial' attach and further re-attach events? I mean, for the initial attach we don't need to add the listener, but should we add the listener upon `setDataProvider` between further reattachments? The logic goes more complex and this is a door for regressions.

Reply:
I think it should work in simply following heurestic:
When setting a data provider
a) if component is attached, initialize everything (most things only take effect on before-client-response phase)
b) if component is not attached, wait for it to be attached before doing everything
On detach things are cleaned up as much as possible, but it may wary (currently) between components whether the data is reused or not (probably not). 
On setting a new data provider, a) and b) rules should always apply.

So I don't at the moment see the reason for the more complex logic between initial and reattach (when it comes to data provider things, browser related things are different).

Reply:
Thanks for clarifications! I'll try to implement this logic.

Reply:
@pleku you approved this PR, so is it fine if I implement this suggestion in the separate PR?

Reply:
Yes I think it should be a separate PR as it is additional and between refactor/bugfix. And another issue should be created as this does handle the original one.

Reply:
And the point about calling `refresh` or `refreshAll` doesn't make sense in my opinion if the component is not attached - since there should not be anything cached (or is there and then why?) then the _refresh_ is a NOOP as all the items will be anyway updated/refetched once the component is attached again.

In case some component retains the fetched items over detach+reattach, that sounds like premature optimization to me.